### ISSUE TITLE
Transform Events Based on Type

### ISSUE THREAD

**Issue by searchy2** [felangel/bloc#524]
How do you transform events only if they match a specific type in `transformEvents`? 
Example:
In `mapEventToState`, you can map to different states based on the event type. Here, there is PostGetData, PostRefresh, and PostLike.
```
  @override
  Stream<PostState> mapEventToState(PostEvent event) async* {
    if (event is PostGetData) {
      yield PostThreadLoading();
      yield* postThreadRefresh();
    }
    if (event is PostRefresh) {
      yield* postRefresh();
    }
    if (event is PostLike) {
      StatusResponse statusResponse =
          await networkProvider.postPostLike(postId, event.like);
      yield PostRefreshing(postThread: postThread);
      dispatch(PostRefresh());
    }
  }
```
I need to debounce the PostLike event to prevent the user from hitting the like button too many times and spamming the server.
In `transformEvents`, you can debounce all events as follows:
```
  @override
  Stream<PostState> transformEvents(
    Stream<PostEvent> events,
    Stream<PostState> Function(PostEvent event) next,
  ) {
      return super.transformEvents(
        (events as Observable<PostEvent>).debounceTime(
          Duration(milliseconds: 500),
        ),
        next,
      );
    }
 }
```
How can I debounce only PostLike events? 
Thanks!

**Comment by AhmedMozaly** [comment]
You can give **whereType** a try
https://pub.dev/documentation/rxdart/latest/rx/Observable/whereType.html

**Comment by searchy2** [comment]
Thanks @AhmedMozaly for taking a look. The problem with whereType and filterBy is that it removes events that I still need. 
My goal is to reach into the stream and only filterBy the specific event I want to debounce. Is that possible with a single stream?

**Comment by felangel** [comment]
Hi @searchy2 👋 
Thanks for opening an issue!

Have you tried something like:

```dart
@override
Stream<CounterState> transformEvents(
  Stream<PostEvent> events,
  Stream<PostState> Function(PostEvent event) next,
) {
    final observableStream = events as Observable<PostEvent>;
    final nonDebounceStream =
      observableStream.where((event) => event is! PostRefresh);
    final debounceStream = observableStream
      .where((event) => event is PostRefresh)
      .debounce(Duration(milliseconds: 300));
    return super.transform(Observable.merge([nonDebounceStream, debounceStream]), next);
}
```

Hope that helps 👍

**Comment by searchy2** [comment]
Thanks for your help! It seems I do need two streams; an extra one for the type I need to filter.

**Comment by craiglabenz** [comment]
For anyone else who comes here in a post-extension methods world, this is the new equivalent:

```dart
@override
Stream<CounterState> transformEvents(
  Stream<PostEvent> events,
  Stream<PostState> Function(PostEvent event) next,
) {
    final nonDebounceStream =
      events.where((event) => event is! PostRefresh);
    final debounceStream = events
      .where((event) => event is PostRefresh)
      .debounceTime(Duration(milliseconds: 300));
    return super.transformEvents(MergeStream([nonDebounceStream, debounceStream]), next);
}
```

**Comment by searchy2** [comment]
A few updates in trying to implement debounce:
- debounce is not part of the default Stream methods. Instead, it can be added with https://pub.dev/packages/stream_transform
- debounced events do not emit until the delay for the last event has been met. This is not the desired behavior as I would like each event to be emitted every x seconds, not until x seconds have elapsed since the last event. The problem is that the events keep coming so an event is never emitted because the delay is not met.

**Comment by felangel** [comment]
@searchy2 `debounceTime` is part of `rxdart` which migrated to extensions so you don't need to use `stream_transform`.

The behavior you're describing doesn't sound like debounce. You might want to take a look at the [rxdart transformers](https://github.com/ReactiveX/rxdart/tree/master/lib/src/transformers) and see what meets your requirements (maybe [throttle](https://github.com/ReactiveX/rxdart/blob/master/lib/src/transformers/backpressure/throttle.dart)?). Ultimately, you have access to the `Stream` and can manipulate it however you'd like. Hope that helps!

**Comment by searchy2** [comment]
Thanks for pointing me in the right direction @felangel 

I wish you a great day and look forward to meeting you at a Flutter meetup some day!

**Comment by synapticvoid** [comment]
Hi,

I just updated @craiglabenz great answer to match the latest `bloc` version (4.0.0):
```dart
  @override
  Stream<Transition<Event, State>> transformEvents(
    Stream<Event> events,
    TransitionFunction<Event, State> transitionFn,
  ) {
    final nonDebounceStream =
        events.where((event) => event is! PostRefresh);

    final debounceStream = events
        .where((event) => event is PostRefresh)
        .debounceTime(Duration(milliseconds: 300));

    return super.transformEvents(
        MergeStream([nonDebounceStream, debounceStream]), transitionFn);
  }
```

Thanks again for this wonderful library!

**Comment by venkateshvk7** [comment]
Hi,

I'm beginner, i'm getting issue with this


  @override
  Stream<LoginState> transformEvents(
    Stream<LoginEvent> events,
    Stream<LoginState> Function(LoginEvent event) next,
  ) {
    final nonDebounceStream = events.where((event) {
      return (event is! EmailChanged || event is! PasswordChanged);
    });

    final debounceStream = events.where((event) {
      return (event is EmailChanged || event is PasswordChanged);
    }).debounceTime(Duration(milliseconds: 300));

    return super.transformEvents(
      nonDebounceStream.mergeWith([debounceStream]),
      next,
    );
  }

@felangel  i tried solve but not solve, can you please help me!
Thank You

**Comment by felangel** [comment]
@venkateshvk7 check out the comment above by @synapticvoid for the updated syntax 👍

**Comment by Mutembeijoe** [comment]
I am well pleased with flutter_bloc.

**Comment by EngAddow** [comment]
> For anyone else who comes here in a post-extension methods world, this is the new equivalent:
> 
> ```
> @override
> Stream<CounterState> transformEvents(
>   Stream<PostEvent> events,
>   Stream<PostState> Function(PostEvent event) next,
> ) {
>     final nonDebounceStream =
>       events.where((event) => event is! PostRefresh);
>     final debounceStream = events
>       .where((event) => event is PostRefresh)
>       .debounceTime(Duration(milliseconds: 300));
>     return super.transformEvents(MergeStream([nonDebounceStream, debounceStream]), next);
> }
> ```

Hi  

> Hi @searchy2 👋
> Thanks for opening an issue!
> 
> Have you tried something like:
> 
> ```
> @override
> Stream<CounterState> transformEvents(
>   Stream<PostEvent> events,
>   Stream<PostState> Function(PostEvent event) next,
> ) {
>     final observableStream = events as Observable<PostEvent>;
>     final nonDebounceStream =
>       observableStream.where((event) => event is! PostRefresh);
>     final debounceStream = observableStream
>       .where((event) => event is PostRefresh)
>       .debounce(Duration(milliseconds: 300));
>     return super.transform(Observable.merge([nonDebounceStream, debounceStream]), next);
> }
> ```
> 
> Hope that helps 👍

Hi @felangel , @craiglabenz and @synapticvoid  I have page with products where user can add to wishlist or remove them and I wanted to not allow the user to click add to wishlist button while I am processing one So I got this post and tried to use all the three examples you gave here guys and it is giving me errors I am using **bloc 6.0.0**  I want to add debounce time to add and remove wishlist events and all other events as normal so pls give me example for my version thanks.

**Comment by valehasadli** [comment]
> ```dart
> nonDebounceStream
> ```

hello, I'm using `flutter_bloc: ^6.1.1` version, 

```dart

  @override
  Stream<Transition<TranslationEvent, TranslationState>> transformEvents(
    Stream<TranslationEvent> events,
    TransitionFunction<TranslationEvent, TranslationState> transitionFn,
  ) {
    final nonDebounceStream = events.where((event) => event is! TranslationFormReset);
    final debounceStream = events
        .where((event) => event is TranslationFormSubmit)
        .debounceTime(const Duration(seconds: 3));

    return super.transformEvents(
      MergeStream([nonDebounceStream, debounceStream]),
      transitionFn,
    );
  }
```

but it doesn't work, debounce time not work in this case.

**Comment by felangel** [comment]
@valehasadli can you please provide a reproduction sample? Thanks!

**Comment by IlyaMax** [comment]
example for version ^8.0.1 
  ```
 transformer: (events, mapper) {
        final debouncedEvents = events
            .where((event) => event.needsDebounce)
            .debounceTime(const Duration(milliseconds: 300));
        final notDebouncedEvents =
            events.where((event) => !event.needsDebounce);
        return MergeStream([debouncedEvents, notDebouncedEvents])
            .flatMap(mapper);
      },
```

**Comment by RaadhukaSubramanian** [comment]
Hi All,

Recently I have upgraded the Bloc version to 8.1.1 from my older version. It says EventTransformer is not deprecated. As I am new to the Flutter development part, I would request you to provide an example how it exactly works.

Below is my code:

Stream<LoginState> transformEvents(
    Stream<LoginEvent> events,
    Stream<LoginState> Function(LoginEvent event) next,
  ) {
    final observableStream = events as Observable<LoginEvent>;
    final nonDebounceStream = observableStream.where((event) {
      return (event is! EmailChanged && event is! PasswordChanged);
    });
    final debounceStream = observableStream.where((event) {
      return (event is EmailChanged || event is PasswordChanged);
    }).debounceTime(Duration(milliseconds: 300));
    return super.transformEvents(nonDebounceStream.mergeWith([debounceStream]), next);
  }
